Source/WebCore/dom/ScriptExecutionContext.cpp

```diff
@@ -851,6 +855,45 @@ CheckedRef<EventLoopTaskGroup> ScriptExecutionContext::checkedEventLoop()
     return eventLoop();
 }
 
+class ScriptExecutionContextDispatcher final
+    : public ThreadSafeRefCounted<ScriptExecutionContextDispatcher>
+    , public RefCountedSerialFunctionDispatcher {
+public:
+    static Ref<ScriptExecutionContextDispatcher> create(ScriptExecutionContext& context) { return adoptRef(*new ScriptExecutionContextDispatcher(context)); }
+
+    // RefCountedSerialFunctionDispatcher
+    void ref() const final { ThreadSafeRefCounted::ref(); }
+    void deref() const final { ThreadSafeRefCounted::deref(); }
+
+private:
+    explicit ScriptExecutionContextDispatcher(ScriptExecutionContext& context)
+        : m_identifier(context.identifier())
+        , m_threadId(context.isWorkerGlobalScope() ? Thread::current().uid() : 1)
+    {
+    }
+
+    // RefCountedSerialFunctionDispatcher
+    void dispatch(Function<void()>&& callback) final
+    {
+        if (m_threadId == 1) {
+            callOnMainThread(WTFMove(callback));
+            return;
+        }
+        ScriptExecutionContext::postTaskTo(m_identifier, WTFMove(callback));
+    }
+    bool isCurrent() const final { return m_threadId == Thread::current().uid(); }
+
+    ScriptExecutionContextIdentifier m_identifier;
+    uint32_t m_threadId { 1 };
+};
+
+RefCountedSerialFunctionDispatcher& ScriptExecutionContext::nativePromiseDispatcher()
+{
+    if (!m_nativePromiseDispatcher)
```

Comment:
should we add threading / current assertion here?

As this code isn't thread-safe but will be used across-threads

Reply:
ScriptExecutionContext::nativePromiseDispatcher is mono thread, it should only be called in context thread.

Reply:
I think just `dispatcher()` would be better; as it has no direct link to NativePromise as such